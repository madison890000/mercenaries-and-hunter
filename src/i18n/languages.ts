import enUS from './locales/en-US.json';
import enTE from './locales/en-TE.json';
import zhCN from './locales/zh-CN.json';
import zhTW from './locales/zh-TW.json';

export type LocalesKey = 'en-US' | 'zh-CN' | 'zh-TW' | 'en-TE';
export const locales: {
    [key in LocalesKey]: Record<string, string>;
} = {
    'en-US': enUS,
    'zh-CN': zhCN,
    'zh-TW': zhTW,
    'en-TE': enTE
};

export const APP_LOCALES: {
    [key in LocalesKey]: string;
} = {
    'en-US': 'English',
    'zh-CN': '中文简体',
    'zh-TW': '中文繁体',
    'en-TE': 'Māori'
};

export const LANGUAGE_OPTIONS = Object.entries<string>(APP_LOCALES).map(([lKey, lValue]) => {
    return {
        label: lValue,
        value: lKey
    } as { label: string; value: LocalesKey };
});
