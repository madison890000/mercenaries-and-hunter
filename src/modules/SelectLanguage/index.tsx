import React from 'react';
import Button from '@mui/material/Button';
import Popper from '@mui/material/Popper';
import Grow from '@mui/material/Grow';
import Paper from '@mui/material/Paper';
import MenuList from '@mui/material/MenuList';
import MenuItem from '@mui/material/MenuItem';

interface SelectLanguageProps {
    value?: string;
    onChange?: (e: string) => void;
    options: { label: string; value: string }[];
}

const SelectLanguage = ({ onChange, value, options }: SelectLanguageProps) => {
    const [open, setOpen] = React.useState(false);
    const anchorRef = React.useRef<HTMLButtonElement>(null);
    const handleToggle = () => {
        setOpen((prevOpen) => !prevOpen);
    };
    const handleClose = (e: string) => {
        onChange?.(e)
        setOpen(false);
    };
    return (
        <>
            <Button
                ref={anchorRef}
                id="composition-button"
                data-testid="select-language-btn"
                aria-controls={open ? 'composition-menu' : undefined}
                aria-expanded={open ? 'true' : undefined}
                aria-haspopup="true"
                onClick={handleToggle}
            >
                {options?.find((l: any) => l.value === value)?.label}
            </Button>
            <Popper
                open={open}
                anchorEl={anchorRef.current}
                role={undefined}
                placement="bottom-start"
                transition
                disablePortal
            >
                {({ TransitionProps, placement }) => (
                    <Grow
                        {...TransitionProps}
                        style={{
                            transformOrigin:
                                placement === 'bottom-start' ? 'left top' : 'left bottom',
                        }}
                    >
                        <Paper>
                            <MenuList
                                autoFocusItem={open}
                                id="composition-menu"
                                data-testid="select-language-menu-pop"
                                aria-labelledby="composition-button"
                            >
                                {options?.map((l: any) => (
                                    <MenuItem onClick={() => {
                                        handleClose(l?.value)
                                    }}>{l?.label}</MenuItem>
                                ))}
                            </MenuList>
                        </Paper>
                    </Grow>
                )}
            </Popper>
        </>
    )
}

export default SelectLanguage
