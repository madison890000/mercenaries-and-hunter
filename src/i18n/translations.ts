import { APP_LOCALES,locales } from './languages';

export const generatedTranslations = () => {
  const translations: any = {};
  for (const key of Object.keys(APP_LOCALES)) {
    try {
      // @ts-ignore
      translations[key] = locales[key];
    } catch {
      console.warn(`Can't find translations for ${key}`);
    }
  }
  return translations;
};
