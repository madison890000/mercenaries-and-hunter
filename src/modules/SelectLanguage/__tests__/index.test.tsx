import React from 'react';
import SelectLanguage from '../index';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { LANGUAGE_OPTIONS } from '../../../i18n/languages';

test('renders SelectLanguage', () => {
    const SelectLanguageDom = renderer.create(<SelectLanguage options={LANGUAGE_OPTIONS} />)
        .toJSON();
    expect(SelectLanguageDom)
        .toMatchSnapshot();
});


test('open SelectLanguage', async () => {
    render(<SelectLanguage options={LANGUAGE_OPTIONS} />);
    const selectBtn = screen.getByTestId('select-language-btn');
    const notShowMenuPops = await screen.queryByTestId('select-language-menu-pop');
    expect(notShowMenuPops)
        .toEqual(null);
    fireEvent.click(selectBtn)
    const menuPops = screen.getByTestId('select-language-menu-pop');
    expect(menuPops)
        .toBeVisible();
})
test('SelectLanguage contain English 中文 中文繁体 中文简体', async () => {
    render(<SelectLanguage options={LANGUAGE_OPTIONS} />);
    const selectBtn = screen.getByTestId('select-language-btn');
    fireEvent.click(selectBtn)
    const menuPops = screen.getByTestId('select-language-menu-pop');
    expect(menuPops)
        .toBeVisible();
    expect(menuPops.innerHTML)
        .toContain('中文简体');
    expect(menuPops.innerHTML)
        .toContain('中文繁体');
    expect(menuPops.innerHTML)
        .toContain('中文');
    expect(menuPops.innerHTML)
        .toContain('English');
})
test('open SelectLanguage snapshot', async () => {
    render(<SelectLanguage options={LANGUAGE_OPTIONS} />);
    const selectBtn = screen.getByTestId('select-language-btn');
    fireEvent.click(selectBtn)
    const menuPops = screen.getByTestId('select-language-menu-pop');
    expect(menuPops)
        .toBeVisible();
    expect(document.body)
        .toMatchSnapshot();
})
test('close SelectLanguage snapshot', async () => {
    render(<SelectLanguage options={LANGUAGE_OPTIONS} />);
    const selectBtn = screen.getByTestId('select-language-btn');
    fireEvent.click(selectBtn)
    const menuPops = screen.queryByTestId('select-language-menu-pop');
    expect(menuPops)
        .toBeVisible();
    fireEvent.click(selectBtn);
    await waitFor(() => {
        expect(screen.queryByTestId('select-language-menu-pop'))
            .toBe(null);
        expect(document.body)
            .toMatchSnapshot();
    })

})
