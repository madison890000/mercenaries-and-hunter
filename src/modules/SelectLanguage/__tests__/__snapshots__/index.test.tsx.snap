// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`close SelectLanguage snapshot 1`] = `
<body>
  <div>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="select-language-btn"
      id="composition-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`open SelectLanguage snapshot 1`] = `
<body>
  <div>
    <button
      aria-controls="composition-menu"
      aria-expanded="true"
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      data-testid="select-language-btn"
      id="composition-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="css-advny6-MuiPopper-root MuiPopperUnstyled-root"
      data-popper-escaped=""
      data-popper-placement="bottom-start"
      data-popper-reference-hidden=""
      style="position: absolute; top: 0px; left: 0px; margin: 0px; transform: translate(0px, 0px);"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
        style="opacity: 1; transform: scale(1, 1); transform-origin: left top; transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <ul
          aria-labelledby="composition-button"
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
          data-testid="select-language-menu-pop"
          id="composition-menu"
          role="menu"
          tabindex="-1"
        >
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            role="menuitem"
            tabindex="0"
          >
            English
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            role="menuitem"
            tabindex="-1"
          >
            中文简体
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            role="menuitem"
            tabindex="-1"
          >
            中文繁体
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <li
            class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
            role="menuitem"
            tabindex="-1"
          >
            中文
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders SelectLanguage 1`] = `
<button
  aria-haspopup="true"
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  id="composition-button"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
/>
`;
